monai/inferers/merger.py

```diff
@@ -0,0 +1,162 @@
+# Copyright (c) MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+from collections.abc import Sequence
+from typing import Any
+
+import torch
+
+from monai.utils import ensure_tuple_size
+
+__all__ = ["Merger", "AvgMerger"]
+
+
+class Merger(ABC):
+    """
+    A base class for merging patches.
+    Extend this class to support operations for `PatchInference`.
+    There are two methods that must be implemented in the concrete classes:
+
+        - aggregate: aggregate the values at their corresponding locations
+        - finalize: perform any final process and return the merged output
+
+    Args:
+        output_shape: the shape of the merged output tensor. Default to None.
+        device: the device where Merger tensors should reside.
+    """
+
+    def __init__(self, output_shape: Sequence[int] | None = None, device: torch.device | str | None = None) -> None:
+        self.output_shape = output_shape
+        self.device = device
+        self.is_finalized = False
+
+    @abstractmethod
+    def aggregate(self, values: torch.Tensor, location: Sequence[int]) -> None:
+        """
+        Aggregate values for merging.
+        This method is being called in a loop and should add values to their corresponding location in the merged output results.
+
+        Args:
+            values: a tensor of shape BCHW[D], representing the values of inference output
+            location: a tuple/list giving the top left location of the patch in the output.
+
+        Raises:
+            NotImplementedError: When the subclass does not override this method.
+
+        """
+        raise NotImplementedError(f"Subclass {self.__class__.__name__} must implement this method.")
```

Comment:
As you already added `@abstractmethod` decorator, this line code will never be executed?

Thanks.

Reply:
It can possibly be called if someone call it from the concrete class using `super()`. I agree that removing it won't cause any harm but the rest of MOANI is like that. Shall we keep it as is and revisit for all the monai base clsses that use both `@abstractmethod` and `raise NotImplementedError`?

https://docs.python.org/3/library/abc.html#abc.abstractmethod
"Note: Unlike Java abstract methods, these abstract methods may have an implementation. This implementation can be called via the `super()` mechanism from the class that overrides it. This could be useful as an end-point for a super-call in a framework that uses cooperative multiple-inheritance."

Reply:
OK, let's leave it as other MONAI modules.

Thanks.